Xml_variable into new_xml

Setnea opened the pull request:
Changes introduce:
- new classes - variable and reference variable
- new xml parsing system

According to new instances parameters and updates parsing system was modified.
Moreover names dict_local and dict_global were replaced by env_local and env_global.

Setnea committed:
New xml params system #120

Setnea committed:
[FIX] log msg fix

Setnea committed:
Updating tasks

Removing parameters/updates_by_name and parameters/updates_by_value by
"pure" parameters and updates.

Setnea committed:
[FIX] TASK_FOR

Setnea committed:
[MINOR] variables name change

Setnea committed:
Adding VariableList class #71

stork119 commented on python/modules/task.py:
```diff
@@ -54,29 +53,30 @@ def parse_task_arguments(self, env_local):
-        logger.debug("Dict_out before concatenation: %s", dict_out.keys())
+    def update_env(self, dict_out, dict_in, variables):
+        for key, var in variables.items():
+            var_class = str(var.__class__.__name__)
```
It's not compatible with object-oriented programming  👎  
You should create function get_variable() in class Variable and overload it in VariableList to get the same functionality.
Remember if you need to check what is a class of the object it means that you use OOP improperly.

Setnea replied:
Right. Changed.

stork119 commented on python/modules/task.py:
```diff
@@ -54,29 +53,30 @@ def parse_task_arguments(self, env_local):
+            else:
+                try:
+                    new_var = var.get_reference(dict_in)
+                except:
```
This exception should be catched in VariableReference class

Setnea replied:
Changed.

stork119 commented on python/modules/xml_parser.py:
```diff
@@ -1,109 +1,143 @@
+
+    def get_reference(self, env): # function name requires changes
+        ref_value = self.get_value(env)
+        var = Variable(self.key, ref_value)
```
No. You do not create new Variable but return Variable from environment

Setnea replied:
I have to do it that, just as we discussed it today.

stork119 reviewed (changes requested):
Move Variables classes into other file. Read comments in code.

Setnea committed:
Merge pull request #124 from stork119/for_task_hotfix

for_task_hotfix merge into xml_variable

Setnea committed:
[FIX] Changed after code review

Setnea committed:
Merge branch 'xml_variable' into xml_variable_list

Setnea committed:
Merge pull request #126 from stork119/xml_variable_list

xml_variable_list into xml_variable

Setnea committed:
[FIX] VariableList

Fixing VariableList according to changes in code review. Removing
unnecesary comments.

Setnea committed:
[FIX] Additional args (from cmd) #133

Setnea committed:
Separate module for Variables #134

Setnea committed:
Task changes #127 part I

Task changes ("standard task": without parallelize, R, FFC)

Setnea committed:
Task changes #127 part II

Changes in parallelize, ffc and r tasks.

Setnea committed:
Adding NEW VariableList #136

Renaming old VariableList to VariableParted. Adding new variable-
VariableList.

Setnea committed:
Merge pull request #135 from stork119/xml_variable_hotfix

xml_variable_hotfix into xml_variable

Setnea committed:
VariableMP #130 part I

Creating placeholder for new class

Setnea committed:
VariableMP #130 #131 part II

Moving some map_plate functions to new class VariableMP

stork119 commented on python/modules/variable.py:
```diff
@@ -0,0 +1,54 @@
+#! /usr/bin/python
```
Why there are not any logs in the functions ?

Setnea replied:
Hm, I wanted to avoid adding too many logs by "just one line" due to all those functions are being used repeatedly during parsing task variables. I'll some logs to VariableParted (before and after concatenation), but I think Var, VarRef i VarList don't require additional msges atm.

stork119 commented on python/modules/task.py:
```diff
@@ -54,69 +54,45 @@ def parse_task_arguments(self, env_local):
-        logger.debug("Dict_out after concatenation: %s", dict_out.keys())        
+    def update_env(self, dict_out, dict_in, variables):
+        for key, var in variables.items():
+            var_out = VAR.get_variable(dict_in)
```
You removed logs from this function but you didn't add them to class Variable

stork119 commented on python/modules/task.py:
```diff
@@ -54,69 +54,45 @@ def parse_task_arguments(self, env_local):
-                value = "".join(value)
-                temp_dict[key_list[i]] = value
-            dict_out = temp_dict.copy()
+    def convert_to_var_dict(self, dict_in):
```
Is it local function ? If so, it should be named _convert_to_var_dict

Setnea replied:
Well it is as local as other task class functions (parse_task_arguments and update_env) except this function is called by many subtasks when update_env is called only by superclass TASK method. Would you like me to convert  all their names ($name$ -> _$name$)? I would rather change parse_task_arguments and update_env names and leave convert_to_var_dict as it is.

stork119 commented on python/modules/task.py:
```diff
@@ -536,13 +534,15 @@ class TASK_READ_DATAFRAME_FROM_CSV(TASK):
         input_path = FM.path_join(dict_setts["input_path"], dict_setts["filename"])
         data = R_connection.read_dataframe_from_csv(input_path, dict_setts["delimiter"])
-        env_local[dict_setts["dict_key_name"]] = data
+        dict_key = dict_setts["dict_key_name"]
```
#145

Setnea replied:
Ok, I'll update this but not on this (new xml) branch.

stork119 commented on python/modules/task.py:
```diff
@@ -551,12 +551,13 @@ class TASK_WRITE_DATAFRAME_TO_CSV(TASK):
     def execute_specify(self, env_local, dict_setts):
         output_path = FM.path_join(dict_setts["output_path"], dict_setts["filename"])
-        data = env_local[dict_setts["dict_key_name"]]
+        v_data = env_local[dict_setts["dict_key_name"]]
```
#145

Setnea replied:
Same as above.

stork119 commented on python/modules/variable.py:
```diff
@@ -0,0 +1,54 @@
+    def get_variable(self, env):
+        return self
+
+class VariableReference(Variable):
```
Add comment how such parameter is parsed in xml

stork119 commented on python/modules/variable.py:
```diff
@@ -0,0 +1,54 @@
+    def get_value(self, env, args = {}):
+        return env[self.value].get_value(env, args)
+
+class VariableList(Variable):
```
Add comment how such parameter is parsed in xml

stork119 commented on python/modules/task.py:
```diff
@@ -474,60 +467,65 @@ class TASK_R(TASK):
+        TASK.__init__(self, parameters, updates, args)
 
     def execute_specify(self, env_local, dict_setts):
         external_params = ["r_function_name", "r_script_path"]
```
These parameters names are defined in dict_task

Setnea replied:
Changed to external_params = list(self.dict_task.keys())

stork119 reviewed (commented):
Look at comments in the code.

Setnea committed:
[MINOR] adding comment to variablelist

Setnea committed:
Merge pull request #142 from stork119/xml_variable_list2

xml_variable_list2 into xml_variable

Setnea committed:
[MINOR] code review

Setnea committed:
[FIX]

Changing VariableReference get_variable() function.
Fixing return in VariableParted get_value() function.

Setnea committed:
Merge branch 'xml_variable_mp' into xml_variable

Setnea committed:
Merge pull request #149 from stork119/xml_variable

Updating xml_variable_mp

Setnea committed:
Adding class VariablePath

Setnea committed:
Functions for VariablePath #138

Setnea committed:
VariablePath modyfications #138 #71

Changing init of VariablePath class.
Modyfing xml_parser according to new class instance.

Setnea committed:
Removing changes from other branches.

Setnea committed:
[MINOR][FIX]

Setnea committed:
[MINOR][FIX]

Setnea committed:
Class VariableMP cd #130

Setnea committed:
VariableMP #130 #131

Setnea committed:
TASK_IF update #132

Updateing task_if to remove map_plate exeption. Changing the way of
parsing map_plate to support parsing reference well names.

Setnea committed:
VariableStructure #147

Setnea committed:
Adding comments to VariableMP

Adding comments describing how map_plate structure and its elements can
be refered in xml settings

Setnea committed:
Changing VariableStructure #147

Changing VariableStructure from list of dictionaries to dictionary.

Setnea committed:
Task FOR update #137

Setnea committed:
TASK_FOR and dependencies modyfication #137

Setnea committed:
TASK_MERGE_SUBDIR_CSV hotfix #148

Setnea committed:
[MINOR] Changes after code review

Moving function from task module to csv_managment.

Setnea committed:
Merge pull request #153 from stork119/xml_variable_hotfix2

hotfix2 into xml_variable

Setnea committed:
Adding comments

Setnea committed:
Merge branch 'xml_variable' into xml_variable_mp

Setnea committed:
Merge pull request #151 from stork119/xml_variable_mp

xml_variable_mp into xml_variable

Setnea committed:
[MINOR] Code review

Setnea committed:
[MINOR] Code review

Setnea committed:
Merge branch 'xml_variable' into xml_variable_path2

Setnea committed:
Merge pull request #150 from stork119/xml_variable_path2

xml_variable_path2 into xml_variable

Setnea committed:
Merge branch 'xml_variable' into xml_variable_struc

Setnea committed:
Merge pull request #152 from stork119/xml_variable_struc

xml_variable_struc into xml_variable

Setnea committed:
[FIX] VariableStructure and VariableList

Changing function get_variable()

Setnea committed:
[MINOR] Adding log msg

Setnea committed:
[BUG] Debug after merges

Setnea committed:
Adding comments to variable module #154

Setnea committed:
[FIX] Fixing VariableMP and others

Setnea committed:
[FIX] Debugging

Setnea committed:
[BUG] VaiableList/Struc and TASK_FOR modyfications

Changing VariableList and VariableStructure. Replacing get_raw_value()
by get_value(), adding get_converted_value() function.
Changing TASK_FOR function and its xml parser.

Setnea committed:
[MINOR] Changing VariableRef

Changing get_variable() function of VariableReference. Given variable
won't be copied to avoid parsing old, original key, instead of previous
method function will create new object of reference class.

Setnea committed:
[MINOR] FFC and R tasks

Editing FFC and R tasks functions to meet new xml formula

Setnea committed:
Add function assign_value

Adding function assign_ref_value, which allows to pass raw reference
value for instances which get_value functions returns sth different than
self.value (i.e. VariableMP objects).

Setnea committed:
Relocating mp related functions

Moving functions related to map_plate from flow_control module to
VariableMP class

Setnea committed:
[BUGFIX] Apply map_plate bugfix

Setnea committed:
[FIX]

duboviy commented on python/modules/variable.py:
```diff
@@ -0,0 +1,381 @@
+            well = self.args["well"].get_value(env)
+            try:
+                self.mp_dict = env[self.value].get_mp_dict()
+            except:
```
- To catch any exception use the ```except Exception```, instead of ```except BaseException``` or just ```except```.
- Indicate the most specific exception type in ```except``` block.

For example:
```python
# Bad
try:
    mapping[key]
except Exception:
    ...

# Better
try:
    mapping[key]
except KeyError:
    ...
```

Setnea replied:
Hey, @duboviy. I added issue to fix all nasty exceptions handling. I'm afraid there are many many more in the whole software code, by now I fixed those called by you. Thanks!

duboviy commented on python/modules/xml_parser.py:
```diff
@@ -1,109 +1,174 @@
+    task_name = name.get('class')
+    try:
+        task_initialization = getattr(TK, task_name)
+    except:
```
Same here

duboviy commented on python/modules/xml_parser.py:
```diff
@@ -1,109 +1,174 @@
+        logger.info("Parsing additional config input_settings.")
+    try:
+        tree = ET.parse(input_path)
+    except:
```
Same here

Setnea committed:
[MINOR] R fisher changes

Setnea committed:
[FIX]

stork119 commented on python/modules/task.py:
```diff
@@ -126,47 +104,20 @@ def execute_specify(self, env_local, dict_setts):
-        TASK.__init__(self, parameters_by_value, parameters_by_name, updates_by_value, updates_by_name, args)
+    dict_task = {"arg_1" : {"required" : True},
+                 "arg_2" : {"required" : True}, 
+                 "comparison" : {"requred" : True}}
```
requred --> required

Setnea replied:
Fixed in all cases.

stork119 commented on python/modules/task.py:
```diff
@@ -202,8 +153,8 @@ class TASK_DOWNLOAD(TASK):
     dict_task = {"input_path" : {"required" : True}, 
                  "output_path" : {"requred" : True}}
```
requred --> required

stork119 replied:
@Setnea check all dict_tasks to find mistakes

Setnea replied:
@stork119  Checked and fixed

stork119 commented on python/modules/task.py:
```diff
@@ -243,26 +194,17 @@ class TASK_MERGE_SUBDIR_CSV(TASK):
-            CSV_M.write_csv(out_path, dict_setts["delimiter"], data) #if we would like to write_csv somewhere...
-
+            try:
+                CSV_M.merge_csv_files(csv_name, main_subdir_list, dict_setts["delimiter"], dict_setts["column_name"], dict_setts["output_path"])
```
Line is to long

stork119 commented on python/modules/task.py:
```diff
@@ -567,7 +528,7 @@ class TASK_MERGE_CSV(TASK):
                  "output_path" : {"required" : True},
                  "delimiter" : {"required" : True, "default" : ","}}
```
input_path_1 -> input_path_2

Setnea replied:
Whole task will change completly soon, but fixed :)

stork119 commented on python/modules/variable.py:
```diff
@@ -0,0 +1,381 @@
+            converted_list.append(new_var)
+        return converted_list
+        
+    def get_converted_value(self, env):
```
Add comment. Is it local/global function ?
In VariableParted it is local function

Setnea replied:
Well, yes, but in VariableParted it is called anyway by get_value() function so there is no need to leave it as global function.

Function get_converted_value of VariableList and VariableStructure is no longer used in other modules, according to few major changes in variable module, but i would like to leave them for a moment in case of future needs. I can comment them by now.

Setnea committed:
[BUGFIX]

Setnea committed:
[MINOR] Adding R script for metadata correction

Setnea committed:
[FIX] Code review changes

stork119 commented:
@Setnea add some small tasks according to our comments
